flink-libraries/flink-table/src/test/scala/org/apache/flink/table/api/scala/stream/sql/SqlITCase.scala

```diff
@@ -293,6 +297,82 @@ class SqlITCase extends StreamingWithStateTestBase {
     assertEquals(expected.sorted, StreamITCase.testResults.sorted)
   }
 
+  @Test
+  def testBoundPartitionedEventTimeWindowWithRow(): Unit = {
+    val env = StreamExecutionEnvironment.getExecutionEnvironment
+    env.setStreamTimeCharacteristic(TimeCharacteristic.EventTime)
+    env.setStateBackend(getStateBackend)
+    val tEnv = TableEnvironment.getTableEnvironment(env)
+    StreamITCase.clear
+
+    val t1 = env.fromCollection(data)
+      .assignTimestampsAndWatermarks(new TimestampWithLatenessWatermark(0))
```

Comment:
I would suggest to implement a utility `SourceFunction` similar to the ones which are defined inline in PR #3386.
The `SourceFunction` could have a `Seq[Either[(Long, Tuple), Long]]` as input, i.e., either tuples with timestamp or a watermark timestamp. 

This would be useful for many more event-time tests.